profile-widget/divecartesianaxis.cpp

```diff
@@ -208,100 +201,142 @@ void DiveCartesianAxis::updateTicks(int animSpeed)
 }
 
-void DiveCartesianAxis::updateLines(int numTicks, double firstPosScreen, double stepScreen, int animSpeed)
+void DiveCartesianAxis::updateLabels(int numTicks, double firstPosScreen, double firstValue, double stepScreen, double stepValue,
+				     int animSpeed, double dataMinOld, double dataMaxOld)
 {
-	for (int i = 0, count = lines.size(); i < count; i++) {
-		double childPos = ((position == Position::Bottom) != inverted) ?
-					firstPosScreen + i * stepScreen :
-					firstPosScreen - i * stepScreen;
+	if (animSpeed <= 0)
+		labels.clear(); // No animation? Simply redo the labels.
+
+	std::vector<Label> newLabels;
+	newLabels.reserve(numTicks);
+	auto actOld = labels.begin();
+	double value = firstValue;
+	for (int i = 0; i < numTicks; i++, value += stepValue) {
+
+		// Check if we already got that label. If we find unused labels, mark them for deletion.
+		// Note: floating point comparisons should be fine owing to our rounding to integers above.
+		for ( ; actOld != labels.end() && actOld->value < value; ++actOld) {
+			double pos = posAtValue(actOld->value);
+			updateLabel(*actOld, 0.0, pos);
+			newLabels.push_back(std::move(*actOld));
+		}
 
-		if (position == Position::Bottom) {
-			// Fix size in case the scene changed
-			QLineF old = lines[i]->line();
-			lines[i]->setLine(old.x1(), old.y1(), old.x1(), old.y1() + rect.height());
-			Animations::moveTo(lines[i], animSpeed, childPos, rect.top());
+		double pos = ((position == Position::Bottom) != inverted) ?
+					 firstPosScreen + i * stepScreen :
+					 firstPosScreen - i * stepScreen;
+		if (actOld != labels.end() && actOld->value == value) {
+			// Update label, but don't delete it
+			updateLabel(*actOld, 1.0, pos);
+			newLabels.push_back(std::move(*actOld));
+			++actOld;
 		} else {
-			// Fix size in case the scene changed
-			QLineF old = lines[i]->line();
-			lines[i]->setLine(old.x1(), old.y1(), old.x1() + rect.width(), old.y1());
-			Animations::moveTo(lines[i], animSpeed, rect.left(), childPos);
+			// Create new label
+			newLabels.push_back(createLabel(value, pos, dataMinOld, dataMaxOld, animSpeed));
 		}
 	}
 
-	// Add the rest of the needed grid lines.
-	for (int i = lines.size(); i < numTicks; i++) {
-		double childPos = ((position == Position::Bottom) != inverted) ?
-					firstPosScreen + i * stepScreen :
-					firstPosScreen - i * stepScreen;
-		DiveLineItem *line = new DiveLineItem(this);
-		line->setPen(gridPen);
-		line->setZValue(0);
-		lines.push_back(line);
-		if (position == Position::Bottom) {
-			line->setLine(0.0, 0.0, 0.0, rect.height());
-			line->setPos(scene.sceneRect().width() + 10, rect.top()); // position it outside of the scene);
-			Animations::moveTo(line, animSpeed, childPos, rect.top());
-		} else {
-			line->setLine(0.0, 0.0, rect.width(), 0.0);
-			line->setPos(rect.left(), scene.sceneRect().height() + 10);
-			Animations::moveTo(line, animSpeed, rect.left(), childPos);
+	// If there are any labels left, mark them for deletion.
```

Comment:
it took me a while to figure out... could you add "mark them for deletion by setting opacityEnd to zero" or something like that? Because I kept staring at this code, trying to figure out what was going on, having not read the comments below about how the Label class works...

Reply:
Hmmmm... It would probably be a good idea to acually delete these objects at the end of the animation. I'll think about that.

PS: I'll be travelling (to France) tomorrow (if they let me in), so I'll probably be less responsive the upcoming week.

Reply:
Added code to actually delete the labels at the end of the animation. And reworded comment. Please check.

Reply:
yes, the code and comments make more sense now. Thanks